Fixes part of #4302 - Introducing a JUnit annotation & rule for overriding Platform Parameters and Feature Flags

Rd4dev opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes part of #4302 
> [Not fixed in this PR: `Maybe: Introducing additional JUnit functionality to help with testing.`]

This PR includes:

### Feature Flags Enhancements:
- Introduced annotations to enable, disable, and reset feature flags.
- Add support for overriding platform parameters.
- Annotations work at both the class and method levels (-) reset annotation.
- Annotations are repeatable.

### Refactoring Test Module Constants:
- Moved test module constants to a dedicated Platform Parameters Constants Module.
- This module is used exclusively for testing purposes.

### Test File Migration:
- Migrated all test files to production usage.
- Removed the Test Platform Parameter Module.
- Refactored all Platform Parameter Providers away from `TestModule`.

### Oppia Test Rule Validation:
- Implemented a regex-based file content validation to enforce the inclusion of `@get:Rule val oppiaTestRule = OppiaTestRule()` in test files.
- Identified (~)298 test files missing the Oppia Test Rule.
- Declared Oppia Test Rule in every test file (-) scripts and few domain modules
- [To Complete] Analyze incompatible test rule declaration files. [mostly script files]

### Wiki Page Updated
https://github.com/oppia/oppia-android/blob/bcc9d4a9a670673a623240876192b44f5ddc750f/wiki/Platform-Parameters-%26-Feature-Flags.md#how-to-write-tests-related-to-platform-parameters

#### Todo:
- Analyze incompatible test rule declaration files. [mostly script files].

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

Rd4dev committed:
Introducing Junit annotations and Test Rules to override Platform parameters

Initial commit, this works with individual overrides while having multiple enable / diasable annotations mess with the overriding and cause the default value to be thrown even with a dummy feature flag name and this commit doesn't address platform parameter value overrides and doesnot check for Oppia Test Rule to be present on all tests.

Rd4dev committed:
Implementation of class level annotation and repeatable annotations for both feature flags and parameter value overrides

Rd4dev committed:
Fix ktlint issues

Rd4dev committed:
Reverting unnecessary changes

Rd4dev committed:
Fix unused import ktlint issue

Rd4dev commented:
@BenHenning @adhiamboperes Would it be possible to provide a draft review or approval on the implementation to ensure things are on the right track before I proceed with the rest? I would like to confirm if the [mentioned ](https://github.com/oppia/oppia-android/issues/4302#:~:text=Updating%20all%20tests%20to%20use%20the%20production%20module%20instead%20of%20the%20test%20or%20a%20custom%20module%2C%20and%20adding%20OppiaTestRule/test%20annotations%20where%20needed%20to%20ensure%20overrides%20are%20still%20happening.)

> "Updating all tests to use the production module instead of the test or a custom module, and adding OppiaTestRule/test annotations where needed to ensure overrides are still happening." 

is done right. If not then will rework based on the feedback.

BenHenning commented:
Reviewed during CLaM meeting today and feedback was left in real-time.

BenHenning commented:
> @BenHenning @adhiamboperes Would it be possible to provide a draft review or approval on the implementation to ensure things are on the right track before I proceed with the rest? I would like to confirm if the [mentioned ](https://github.com/oppia/oppia-android/issues/4302#:~:text=Updating%20all%20tests%20to%20use%20the%20production%20module%20instead%20of%20the%20test%20or%20a%20custom%20module%2C%20and%20adding%20OppiaTestRule/test%20annotations%20where%20needed%20to%20ensure%20overrides%20are%20still%20happening.)
> 
> > "Updating all tests to use the production module instead of the test or a custom module, and adding OppiaTestRule/test annotations where needed to ensure overrides are still happening."
> 
> is done right. If not then will rework based on the feedback.

Everything looks right except the overriding mechanism. Rather than having an override built into the production module, we should be overriding the value via OkHttp and the platform parameter syncing. This might be a bit tricky to set up, so it may well be worth getting this working before migrating all uses over to the new system.

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into annotations_test_rules_platform_parameter

Rd4dev committed:
Experimenting with enums and SpotlightFragmentTest

Rd4dev committed:
Save lobby for migrating from TestPlatformParameterModule to PlatformParameterModule 1

on hold: TopicRevisionTabSpotlight yet to properly migrate 6 more test cases here

Rd4dev committed:
SharedTests Migration from TestPlatformParamaterModule to PlatformParameterModule

Rd4dev committed:
Done with app module test migrations

Rd4dev committed:
Save lobby for domain tests migration

Gradle tests couldn't resolve dagger test components, so they need to be double checked on the CI and through bazel tests.

Rd4dev committed:
Fix ktlint issues

Rd4dev committed:
Resolved merge conflicts

Rd4dev commented:
Removing stale label, will work on this next week.

Rd4dev committed:
Domain Module test migration

Todo Later: (-minus) ApplicationLifecycleObserverTest, PlatformParameterModuleTest, SurveyGatingControllerTest

Rd4dev committed:
Fix Lint Issues

Rd4dev committed:
Test migration minus 6 test files

PlatformParameterModuleTest, SurveyGatingControllerTest, ApplicationLifecycleObserverTest, SyncStatusManagerImplTest, LogReportingModuleTest, PerformanceMetricsAssessorImplTest

Rd4dev committed:
Fix Lint checks

Rd4dev committed:
Resolve merge conflicts

Rd4dev committed:
Fix StateFragmentTest and ProfileChooserFragment

Running them on CI

Rd4dev committed:
Fix StateFragmentTest and ProfileChooserFragment

Running them on CI

Rd4dev committed:
Reverting latex platform parameter constant

Set to false to test override of parameter value

Rd4dev committed:
Fix HomeActivityTest

Removed the set up for onboarding V2 flag as both disable and enable were set to enable the v2 flag while none of the activation or change of default values made any effect.

Rd4dev committed:
Fix ExplorationActivityTest and reverted ExplorationProgressModuleTest and FeatureFlagsLoggerTest back to TestPlatformParameterModule

Few test files use TestPlatformParameterModuleConstant values which was provided then by the TestPlatformParameterModule but those were just for testing purposes and not sure how to migrate them to production module since that would involve mixing both, either they need to be merged to be in one or handled as a separate file for constants just for testing and injected as required (also to be placed at a more appropriate location if required). Todo: 1. Migration of the left 6 files + these 2 -- 2. To remove TestPlatformParameterModule -- 3. To check for OppiaTestRule instance at every test file

Rd4dev committed:
Migrated all modules from TestPlatformParameterModule to PlatformParameterModule

Todo: 1. Test Module's build file to be refactored, 2. Platform Parameter Overrides, 3. Scan for Oppia Test Rule insertions 4. Remove Test module after (1) and confirm with CI.

Rd4dev committed:
Removed the test platform parameter module

Rd4dev committed:
Trying out Test Rule declaration check with RegexPatternValidationCheck

Testing it via CI as local run threw OutOfMemoryError - Java heap space Exception.

Rd4dev committed:
Added missing OppiaTestRule in every test file

The Regex Validation returned around 298 test files with a missing oppia test rule and all these files were now included with the rule irrespective of the feature flag usage.

Rd4dev committed:
Fix Oppia Test Rule imports in few utility modules

Needs //testing dependency and throws Fingerprint issues without RunWith AndroidJunit4 class.

Rd4dev committed:
Lint fixes and a few more dep additions

Rd4dev committed:
Removed the inclusion of oppia test rule from scripts and fixed few more deps

Rd4dev committed:
Fixing the rest of the test cases to get past unit tests

The parameterized tests do not accept the oppia test rule, similar to the scripts scenario. Todo: The RegexPatternValidationChecks need to be added, To add the kDocs, Override Platform Parameter Values -- to figure out oppia test rule for excluded tests and if needed to exempt them - TodoCheckTest is messed up (haven't looked into it), The RegexPatternValidationCheckTest has few hot fixes that need to be properly validated later that includes adding the oppia test rule for parametrized test case (though production files don't accept them for now) and few log file numbers were modified to accept to reflect the changes (need to validate them too).

Rd4dev committed:
Fix failing instrumentation tests

Rd4dev committed:
Adding a fallback value in getCurrentPlatform

This need to be tested and I'm not sure of this since the scripts assertion throw gets interrupted on the catch calls and mess the test cases so this commit will probably be somthing to verify the workaround in the CI, if the fallback or other workarounds don't work then the scripts module test files in particular might need to be exempted.

Rd4dev committed:
Fix parameterized tests with rule declaration fall back in domain module

The ExplorationPlayerTest in the instrumentation module requires some system images and fails to run locally, I believe other than that only the scripts remain as the other utility module test with parameterization fallback did successfully pass. Todo: 1. Analyze other ways to workaround the fingerprint unavailability 2. To work on override platform parameter values 3. Wiki update 4. KDoc 5. Exempted test files 6. Remove debug logs (printstream).

Rd4dev committed:
Added KDocs for annotations and parameter constants

Yet to add for override platform parameters.

Rd4dev committed:
Lint and KDoc fixes

Rd4dev committed:
Implementation of platform parameter overrides

Unsure of string type parameters and int type parameters were untested.

Rd4dev committed:
Added RegexPatternValidationCheckTests for missing oppia test rule declarations.

Rd4dev committed:
Lint fixes for max line length

Rd4dev committed:
Added PlatformParameterModuleTests, removed parameters provided from TestModule, updated  wiki page

Rd4dev committed:
Added test cases for PlatformParameterModule reset / clear overrides

Rd4dev committed:
Reverting log stream setting

Rd4dev committed:
Fix failing test case that had remaining test module declarations

Rd4dev commented:
@BenHenning, @adhiamboperes PTAL, and I also wanted to confirm a few requirements:

1.  What should be the scope of `OppiaTestRule` inclusion?

    Is it required in every module's test file, or are there exceptions?

    The concern is that `OppiaTestRule` attempts to acquire `Build.FINGERPRINT` to determine the Test Platform. While this works for other modules that access Android classes, the script module test cases throw a `NoClassDefFoundError`. Although the issue can be mitigated by redirecting to use Robolectric as the Test Platform, the previous error logs interfere with current assertion throw cases, leading to test failures. I wanted to clarify the scope and requirements for including the rule before investigating other workarounds.

2. Codeowners Check Failure

    Are there any specific changes I need to address to resolve the [CODEOWNERS check failure](https://github.com/oppia/oppia-android/actions/runs/12678024040/job/35334689300?pr=5565#step:4:16)?

3. Reset Annotation

    As discussed in a prior CLaM meeting, it was suggested to remove the `ResetPlatformParameter` annotation since `overridePlatformParameter` was deemed sufficient. I just have few questions:

    - If `ResetPlatformParameter` was removed for platform parameters, should it also be removed for feature flags? (I remember it being mentioned to not to) but the same functionality (resetting to a default value) could be achieved by utilizing `EnableFeatureFlag` or `DisableFeatureFlag`.
    - If keeping a reset annotation is advisable, would maintaining it for both feature flags and platform parameters be helpful to ensure parity and avoid confusion?
    - Alternatively, would keeping the reset annotation exclusively for feature flags suffice?

Rd4dev commented on testing/src/main/java/org/oppia/android/testing/OppiaTestRule.kt:
```diff
@@ -133,5 +234,34 @@ class OppiaTestRule : TestRule {
+    /**
+     * Extracts all feature flag annotations of the specified type from a collection of annotations.
+     *
+     * @param annotations a collection of annotations to be checked for feature flag annotations.
```
Todo: To remove ending periods from all KDocs.
```suggestion
     * @param annotations a collection of annotations to be checked for feature flag annotations
```

adhiamboperes commented:
Thanks @Rd4dev! Reviewing this today.

Rd4dev commented:
> Thanks @Rd4dev! Reviewing this today.

Thanks @adhiamboperes, appreciate the update!

BenHenning commented on testing/src/main/java/org/oppia/android/testing/EnableFeatureFlag.kt:
```diff
@@ -0,0 +1,14 @@
+ */
+@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION)
+@Repeatable
+annotation class EnableFeatureFlag(val name: FeatureFlag)
```
For both this and the disable variant, it would be good to explain here how cascading cases or repeated cases are handled. For instance, consider the following setup:

```kotlin
@EnableFeatureFlag(FEATURE_1)
@DisableFeatureFlag(FEATURE_1)
class MyConfusingTest {

  @Test
  fun test_case1() {}

  @Test
  @EnableFeatureFlag(FEATURE_1)
  fun test_case2() {}

  @Test
  @DisableFeatureFlag(FEATURE_1)
  fun test_case3() {}

  @Test
  @EnableFeatureFlag(FEATURE_1)
  @DisableFeatureFlag(FEATURE_1)
  fun test_case4() {}

  @Test
  @DisableFeatureFlag(FEATURE_1)
  @EnableFeatureFlag(FEATURE_1)
  fun test_case5() {}
}
```

What is the expected state of ``FEATURE_1`` in each test case? Is it an error or does latest win?

Personally, I would expect that:
- More than one declaration for a single feature at the class or method level results in an error (i.e. ``MyConfusingTest``, ``test_case4``, and ``test_case5`` above should all fail).
- It's allowed to have at most 2 overrides for a single feature for a given test: one at the class level, and one at the method level, where the method annotation always takes precedence. This allows us to enable/disable a feature for the whole class, then selectively enable/disable it on a per-test basis as needed.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/OverrideStringParameter.kt:
```diff
@@ -0,0 +1,14 @@
+ */
+@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION)
+@Repeatable
+annotation class OverrideStringParameter(val name: PlatformParameter, val value: String)
```
For this & the other override classes: see my comment in ``EnableFeatureFlag`` as I think the same consideration is needed here on how cascading and overriding scenarios are handled.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/ResetFeatureFlagToDefault.kt:
```diff
@@ -0,0 +1,14 @@
+ */
+@Target(AnnotationTarget.FUNCTION)
+@Repeatable
+annotation class ResetFeatureFlagToDefault(val name: FeatureFlag)
```
Similar to my comment in ``EnableFeatureFlag``, cascading/duplication needs to be considered here. I'd expect any cases where an enable/disable happens and reset at the same level to be an error (i.e. a test method can only reset, not enable/disable and then reset).

For strictness, I'd also expect that reset throws an error if there's no explicit override of the feature at the class level (i.e. there's no reason to reset). While resetting a non-overridden flag is a no-op, allowing it to be checked in could result in false assumptions and a fundamental misunderstanding of what the test and surrounding test suite is actually doing.

BenHenning reviewed (commented):
Thanks @Rd4dev! This is really exciting to see progress on.

I only looked at the top-level override classes for now, and still need to dig deeper into the PR. However, I do have one thought that I wanted to run by you.

I've had an idea for some time to leverage code generation for the feature flags to minimize the amount of Dagger binding boilerplate that needs to be done each time a new flag is added. Originally, I thought that this would require a custom annotation processor. However, now that Gradle is gone, we actually could do this with protobuf. Consider:
- A new textproto file that lists every feature flag and platform parameter with server string names and default values. I suspect this would be a twofold definition: explicit parameters in a proto file, and server names and default values in textproto (so that we have strong typing for the separate flags).
- Automatic compilation to binary proto (like we do with the language configuration under ``config/``).
- Some file loading logic similar to that which is needed for the languages configuration.
- Updating Dagger bindings that just generate the entire set of bound flags/parameters by loading the textproto file.\*

\* Note that this is a little trickier than it sounds because we can't technically do file I/O within Dagger providers without incurring some performance issues. However, something that we _could_ do is automatically bind implementation classes based on the proto structure (which is available at compile time since the individual flags would be defined within a proto file), and then have the implementations support asynchronously loading their state (either from the local file cache or from their local file configuration). It's possible we don't even need this second bit since I think the platform parameter system is technically supposed to be designed to fully load its asynchronous state during the app startup so that all flag values are fixed for the lifetime of the app process.

While a long explanation, I suspect that the delta might be relatively small and also be a pretty substantial improvement to how flags and parameters are interacted with in both tests and production code. There's probably more that could be done with real code generation, but that starts getting a bit scary in scope here.

Is this something you'd want to take on as part of this PR? If not, no worries. I just wanted to ask since you've already spent a lot of time thinking about the mechanisms. :)

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/home/HomeActivityTest.kt:
```diff
@@ -1985,7 +1916,6 @@ class HomeActivityTest {
 
   @Test
   fun testHomeActivity_onBackPressed_soleLearnerProfile_exitsApp() {
-    setUpTestWithOnboardingV2Enabled()
```
This test and 3 below require the feature flag to be enabled, but they are missing the annotation. Do they pass?

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt:
```diff
@@ -324,8 +338,10 @@ class TopicFragmentTest {
 
   @Test
+  @EnableFeatureFlag(FeatureFlag.SPOTLIGHT_UI)
+  @EnableFeatureFlag(FeatureFlag.EXTRA_TOPIC_TABS_UI)
```
Should this be disabled?

adhiamboperes commented on domain/src/test/java/org/oppia/android/domain/survey/SurveyGatingControllerTest.kt:
```diff
@@ -280,6 +286,7 @@ class SurveyGatingControllerTest {
   }
 
   @Test
+  @EnableFeatureFlag(FeatureFlag.ENABLE_NPS_SURVEY)
```
Since this test class is valid  only for the feature flag enabled situation, I think it is sufficient to enable the flag at class level.

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt:
```diff
@@ -103,31 +113,39 @@ class PlatformParameterModule {
-    ) ?: PlatformParameterValue.createDefaultParameter(
-      ENABLE_EDIT_ACCOUNTS_OPTIONS_UI_DEFAULT_VALUE
-    )
+    return overriddenFeatureFlags[FeatureFlag.EDIT_ACCOUNTS_OPTIONS_UI]?.let {
+      PlatformParameterValue.createDefaultParameter(it)
+    } ?: platformParameterSingleton.getBooleanPlatformParameter(EDIT_ACCOUNTS_OPTIONS_UI)
+      ?: PlatformParameterValue.createDefaultParameter(
+        ENABLE_EDIT_ACCOUNTS_OPTIONS_UI_DEFAULT_VALUE
+      )
```
I don't understand why this is a combination of feature flag and platform parameter here and below.

adhiamboperes reviewed (changes requested):
Thanks @Rd4dev! I have left a few comments, PTAL.

Rd4dev committed:
Resolving merge conflicts -- setup

Rd4dev committed:
Separating the class and method feature flag annotation extractions

The earlier implementation did 2 things 1. retrieved the annotations from both class and method levels and accumulated into one variable 2. it first sets the enable value followed by the disable value. To make sure the method annotation takes precedence the separation of class level and method level annotation retrievals are done and class levels and called first then goes the method levels, but still if there are multiple annotations for the same feature flag made this would always take the disable values for the final consideration which shouldn't be the case, so this need fix to just have one annotation for each class and method max 2 for a single feature flag / platform parameter; also this commit is just a temp test done with just the enable and disable feature flags which will need to be applied for platform parameters too (might need separate helper if gets too messy).

Rd4dev committed:
Validating Conflicting annotation of feature flags

By this a test class / method can have at most one annotation per feature flag - one of enabled, disabled or reset. Things need to be implemented for overridding platform parameters too.

Rd4dev committed:
Cleaning up OppiaTestRule and reverting Test Class

The overridding calls are separated to their own methods. Need classes to properly test platform parameter values especially for int / string values.

Rd4dev committed:
Separating Platform Parameter overrides to class and method levels

Rd4dev committed:
Validation for reset configuration to have an explicit class level annotation

Rd4dev committed:
Disabling EXTRA TOPIC TABS UI feature flag to be in sync with the existing setup

Rd4dev committed:
Fix Lint issues
